engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java

```diff
@@ -293,10 +295,14 @@ public void shouldUseTheSameLoopVariablesWhenIncidentResolved() {
             .create();
 
     // when
-    RecordingExporter.incidentRecords(IncidentIntent.CREATED)
-        .withProcessInstanceKey(processInstanceKey)
-        .limit(3)
-        .map(Record::getKey)
+    final List<Long> incidents =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .limit(3)
+            .map(Record::getKey)
+            .collect(Collectors.toList());
+    ENGINE.variables().ofScope(processInstanceKey).withDocument(Map.of("y", 1)).update();
+    incidents.stream()
         .map(key -> ENGINE.incident().ofInstance(processInstanceKey).withKey(key))
         .forEach(ResolveIncidentClient::resolve);
```

Comment:
This does not look deterministic, although I might be wrong here.

Please also verify that the variable was updated before resolving the incidents.

Reply:
The variable is updated as part of the command processing, as in, the state is modified. `update()` returns the response from the client, so I would assume it's deterministic, as in the variable is 100% updated when we receive the response. See the success expectation:

```java
private static final LongFunction<Record<VariableDocumentRecordValue>>
      SUCCESSFUL_EXPECTATION_SUPPLIER =
          (sourceRecordPosition) ->
              RecordingExporter.variableDocumentRecords(VariableDocumentIntent.UPDATED)
                  .withSourceRecordPosition(sourceRecordPosition)
                  .getFirst();
```

And the update method:

```java
public Record<VariableDocumentRecordValue> update() {
    final long position =
        environmentRule.writeCommand(VariableDocumentIntent.UPDATE, variableDocumentRecord);
    return expectation.apply(position);
  }
```

I simply ignore the return value here, but I think we can safely say the variable was updated. Here are the records produced by updating a document:

```
C VAR_DOCUMENT       UPDATE         - #040->  -1   -1 - VariableDocumentRecord {"variables":{"y":1},"updateSemantics":"PROPAGATE","scopeKey":2251799813685253}
E VAR                CREATED        - #041->#040 K023 - y->1 in <process [K005]>
E VAR_DOCUMENT       UPDATED        - #042->#040 K024 - VariableDocumentRecord {"variables":{"y":1},"updateSemantics":"PROPAGATE","scopeKey":2251799813685253}
```

Reply:
I was unaware that we wait for the response. This makes sense. Thanks for the clear description :)